code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go

```diff
@@ -57,73 +58,63 @@ func challengeTableToStorageChallengeInfo(ch *event.Challenge, edb *event.EventD
 	}, nil
 }
 
-func emitAddChallenge(ch *StorageChallengeResponse, expiredCountMap map[string]int, expiredN int, balances cstate.StateContextI) {
+func emitAddChallenge(ch *StorageChallengeResponse, expiredCountMap map[string]int, expiredN int, balances cstate.StateContextI, allocStats, blobberStats *StorageAllocationStats) {
 	balances.EmitEvent(event.TypeStats, event.TagAddChallenge, ch.ID, storageChallengeToChallengeTable(ch, expiredN))
 	balances.EmitEvent(event.TypeStats, event.TagAddChallengeToAllocation, ch.AllocationID, event.Allocation{
-		AllocationID:    ch.AllocationID,
-		OpenChallenges:  int64(1 - expiredN), // increase one challenge and remove expired ones
-		TotalChallenges: int64(1),
+		AllocationID:         ch.AllocationID,
+		OpenChallenges:       allocStats.OpenChallenges,
+		TotalChallenges:      allocStats.TotalChallenges,
+		SuccessfulChallenges: allocStats.SuccessChallenges,
+		FailedChallenges:     allocStats.FailedChallenges,
 	})
 
-	chBlobberOpenDelta := 1
-	if exp, ok := expiredCountMap[ch.BlobberID]; ok {
-		chBlobberOpenDelta -= exp
-	}
-
 	// Update open challenges count of challenge blobber
-	balances.EmitEvent(event.TypeStats, event.TagUpdateBlobberOpenChallenges, ch.BlobberID, event.ChallengeStatsDeltas{
-		Id:        ch.BlobberID,
-		OpenDelta: int64(chBlobberOpenDelta),
+	balances.EmitEvent(event.TypeStats, event.TagUpdateBlobberOpenChallenges, ch.BlobberID, event.Blobber{
+		Provider:            event.Provider{ID: ch.BlobberID},
+		OpenChallenges:      uint64(blobberStats.OpenChallenges),
+		ChallengesCompleted: uint64(blobberStats.TotalChallenges),
+		ChallengesPassed:    uint64(blobberStats.SuccessChallenges),
 	})
 
-	// Remove expired challenge count of the other blobbers
-	for bid, expiredCount := range expiredCountMap {
-		if bid != ch.BlobberID {
-			// Emit event per blobber and the merger will be able to merge them for each blobber
-			balances.EmitEvent(event.TypeStats, event.TagUpdateBlobberOpenChallenges, bid, event.ChallengeStatsDeltas{
-				Id:        bid,
-				OpenDelta: int64(-expiredCount),
-			})
-		}
-	}
 	logging.Logger.Debug("emitted add_challenge")
 }
 
-func emitUpdateChallenge(sc *StorageChallenge, passed bool, balances cstate.StateContextI) {
+func emitUpdateChallenge(sc *StorageChallenge, passed bool, balances cstate.StateContextI, allocStats, blobberStats *StorageAllocationStats) {
 	clg := event.Challenge{
 		ChallengeID:    sc.ID,
 		AllocationID:   sc.AllocationID,
 		BlobberID:      sc.BlobberID,
-		Responded:      sc.Responded,
 		RoundResponded: balances.GetBlock().Round,
 		Passed:         passed,
 	}
+	if passed {
+		clg.Responded = int64(1) // Passed challenge
+	} else {
+		clg.Responded = int64(2) // Failed challenge
```

Comment:
why 2?

Reply:
For expired it is 2, for open it is 0 and for passed it is 1